Optimize 'affine' handlers

dolio opened the pull request:
This work allows certain handlers to be optimized to avoid manipulating continuations. Right now the handlers detected have some limitations, but more should be possible in the future.

The optimization is centered around the idea of an "affine" handler. This is a self-recursive handler where _every effect case_ falls into one of two forms:

1. A "linear" case, where the continuation is used exactly once, to resume in tail position.
2. An "irrelevant" case, where the continuation is never used.

It's not necessary that the effects are all linear or all irrelevant, a mix is allowed. Also, it's possible to do some pattern matching before it's decided which of the above cases we're in, so something like the following should be optimized:

```unison
h = cases
  { eff n -> k } ->
    match n with
      0 -> 3
      n -> handle k 7 with h
```

The translation tries to figure out what's going on and avoid doing extra handler work unless it's necessary.

The code interchange format is unaffected. The optimized affine handlers are added during code _loading_. So, hashes should be the same, and in a client-server setup, code generated on old clients can benefit from the server having the optimization. Or, code loaded from a database (say) should automatically benefit if the handlers are of the right form.

Performance for `@pchiusano/misc-benchmarks` is largely the same. Here are the cases that differ:

```
fib2
2.887309ms -> 486.425µs

Json parsing (per document)
268.653µs -> 307.115µs
```

I'm not sure why the JSON benchmark is marginally slower. It seems to be a real effect, not just variance. Maybe handlers that aren't optimized have a little more overhead. Maybe I should just investigate it separately, though.

Base tests and transcripts pass.

dolio committed:
Improve ANF pretty printing

Indent some things better.

Use shortened hashes for displaying references, instead of just `show`ing them.

dolio committed:
Add an affine environment to the interpreter

Currently this is unused, and only the existing dynamic handler
environment is ever populated.

dolio committed:
Add machinery for affine handlers to intermediate code and interpreter

dolio committed:
Additional interpreter features for affine handlers

dolio committed:
Roll back immediate affine handler compilation

This keeps the affine handler spot in the intermediate code
representation, but it is not generated during compilation, and is
omitted from the interchange format. The idea is to generate the
affine handlers as a post-process around the time inlining happens.
This should make it easier to recognize affine handlers, and also
means that code interchange, hashes, etc. are unaffected by this
optimization.

dolio committed:
Add a debugging combinator for sequences of definitions with references

dolio committed:
Add an interpreter instruction for entering a local affine scope

This saves the handlers on the stack while the affine handler is
evaluated with its own handlers

dolio committed:
Fix extendPAp

The extra argument was being placed on the wrong end of the segment.

dolio committed:
Add ANF support for affine handlers

These are not generated yet, but they are the machinery necessary for
the optimized handlers. As a first step, the top level exception
handler has been changed to an affine handler.

These constructors are also not part of the serialization format. They
will be generated when the code is loaded so that hashes etc. do not
need to change to support affine handlers.

dolio committed:
Allow inlining into by-name bindings

This will help reduce the analysis necessary for affine handlers.

dolio committed:
Move affine handler translation to a later optimization pass

Allows the optimization to be performed server side for clients that
don't know about it in the near term, and also keeps the interchange
format the same. Also made the translation easier.

dolio committed:
Merge remote-tracking branch 'origin/trunk' into topic/affine-abilities

dolio committed:
Fix performance regression from new affine code

Seems likely to have been from the more complicated `resolve`, but
added some other cleanup as well.

dolio committed:
Allow branching in affine handlers

dolio committed:
Recognize 'irrelevant' handlers and add a little flexibility

dolio committed:
Refine strictness of HEnv to avoid some performance overhead

dolio committed:
Fix exception call for builtins

Was only looking in the denv, but now it can be in the aenv

dolio committed:
Fix unison-runtime tests

dolio commented:
BTW, work on this was a little disjointed. At some point I made a commit just to push things to github (just in case). But the code wasn't exactly in a coherent state at that point. I did some rebasing, but let me know if you see any debugging type stuff that I missed removing.

pchiusano reviewed (approved):
Awesome. This seems like it'll be significant for real code, given that we're already getting 6x speed boost from the little fib examples.

I'd look into the JSON parsing thing. My concern is that that benchmark just uses `Throw`, and it doesn't even hit the `throw` case in practice. So I wonder why it's slower.

I'd also try running the cloud client tests.

aryairani commented:
Go head and merge this whenever you guys are happy with it.

ceedubs commented:
I've verified that the Unison Cloud integration tests pass with this change :+1:

dolio committed:
Don't force CCache in prim1

dolio committed:
Avoid unnecessary evaluation of handler environments

The handler state is not used everywhere, so it's not really
advantageous to keep it evaluated so that the `HEnv` value is unpacked
as arguments to `eval`, `exec`, etc. Instead, the values are forced
when built (to avoid excessive thunking), and inspections of them are
performed only inside cases where they're necessary, to avoid needless
forcing operations.

dolio commented:
I added some performance tuning changes.

Basically, my idea was to _not_ make the handler environments strict arguments to everything. This means that `eval` and such have fewer arguments in the core representation, which seems to help.

Instead, the handler environments are immediately evaluated when built, to avoid thunks building up. I also tweaked the way they're used in various spots so that the code doesn't waste time re-forcing them all over the place.

I also cleaned up some other forcing that was causing some sub-optimal behavior.

The json benchmark _still_ seems a little slower, but it's like 3-5% now. And _everything else_ seems to be slightly faster than it was. So maybe there's still some performance on the table for the json benchmark, but I don't actually know _what_ is slower in that test. Everything it's doing seems to have an analogue in another benchmark, and they're all the same or faster than they were, I think.

dolio committed:
Add some notes about non-strict (h)env values

pchiusano commented:
Note, windows JIT build failed spuriously, but I merged anyway.
